Mathlib/Algebra/FreeMonoid/Basic.lean

```diff
@@ -346,11 +469,82 @@ theorem map_id : map (@id α) = MonoidHom.id (FreeMonoid α) := hom_eq fun _ ↦
 #align free_monoid.map_id FreeMonoid.map_id
 #align free_add_monoid.map_id FreeAddMonoid.map_id
 
+theorem map_surjective (f : α → β) : Function.Surjective f → Function.Surjective (map f) := by
+  intro fs d
+  induction' d using FreeMonoid.inductionOn' with head tail ih
+  · use 1
+    rfl
+  specialize fs head
+  rcases fs with ⟨thingie, rfl⟩
+  rcases ih with ⟨another, rfl⟩
+  use FreeMonoid.of thingie * another
+  rfl
+
+end Map
+
+/-! ### reverse -/
+
+/-- reverses the symbols in a free monoid element -/
+@[to_additive "reverses the symbols in an additive free monoid element"]
+def reverse : FreeMonoid α → FreeMonoid α := List.reverse
+
+@[to_additive (attr := simp)]
+theorem reverse_mul {a b : FreeMonoid α} : reverse (a * b) = reverse b * reverse a :=
+  List.reverse_append _ _
+
+
+/-! ### foldr -/
+
+/--
+Applies function `f` to all of the elements of the free monoid, from right to left.
+* `foldr f init (of a * of b * of c) = f a <| f b <| f c <| init
+-/
+@[to_additive (attr := specialize) "Applies function `f` to all of the elements of the free monoid,
+from right to left. `foldr f init (of a + of b + of c) = f a <| f b <| f c <| init"]
+def foldr (f : α → α → α) (init : α) (a : FreeMonoid α) := List.foldr f init a
```

Comment:
I'm curious what the use case of this would be.

Reply:
It might be overkill to use foldr; perhaps there's a more specific way. In a later file, I'll need to embed into a group. So I want to multiply together all the symbols in the free monoid element, using the group's multiplication

Reply:
Sorry for the late response, but that is just [FreeMonoid.lift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/FreeMonoid/Basic.html#FreeMonoid.lift). The reason, I am suspicious of `foldr` is that I don't think we should just duplicate the `List` API in `FreeMonoid`.

Reply:
oh, I see! that makes sense, I'll remove it

Reply:
What's the status of this?

Reply:
it's deleted as of the last commit